File: python-webservice-template/persistence/stream_store/lsm_store.py
"""In-memory LSM (Log-Structured Merge) store for Arrow record batches.

SINGLE-WRITER CONTRACT — LSMStore has no internal locking. It is designed
to be written by exactly one thread (the ingest consumer thread started by
StreamIngestService). Concurrent calls to `ingest()` produce undefined
behaviour. Readers (`query()`) are safe to call from any thread or coroutine
because `_publish()` swaps the snapshot reference atomically at the end of
every write; readers always see a consistent, immutable snapshot.

Lifecycle: memtable → run (flush) → compacted run.  When the number of runs
reaches `compaction_runs`, all runs are merged into one. Compaction is the
only point at which delete tombstones are reclaimed; a tombstone in the
memtable or an un-compacted run continues to shadow older versions of the
same key across all in-memory frames until a full merge has occurred.
"""
from dataclasses import dataclass

import pyarrow as pa
import polars as pl

ORDER_COLUMN = "seqno"


def _merge_frame(frames: tuple[pl.DataFrame, ...],
                 key_columns: list[str]) -> pl.DataFrame | None:
    if not frames:
        return None
    combined = pl.concat(frames, how="vertical")
    # Window function: rank rows within each key partition by recency
    # (newest seqno first). key_columns is the single extension point:
    # ["id"] today, ["id", "version"] later, with no other change.
    winners = (
        combined
        .with_columns(
            pl.col(ORDER_COLUMN)
            .rank("ordinal", descending=True)
            .over(key_columns)
            .alias("_rn")
        )
        .filter(pl.col("_rn") == 1)
        .drop("_rn")
    )
    return winners


def _merge_to_rows(frames: tuple[pl.DataFrame, ...],
                   key_columns: list[str],
                   limit: int | None) -> tuple[list[dict], int]:
    winners = _merge_frame(frames, key_columns)
    if winners is None:
        return [], 0
    live = winners.filter(pl.col("op") != "delete").sort(key_columns)
    total = live.height
    if limit:
        live = live.head(limit)
    return live.drop([ORDER_COLUMN, "op"]).to_dicts(), total


@dataclass(frozen=True)
class _Snapshot:
    runs: tuple[pl.DataFrame, ...]
    memtable: tuple[pl.DataFrame, ...]


class LSMStore:
    def __init__(self, flush_rows: int, compaction_runs: int,
                 key_columns: list[str] | None = None) -> None:
        self._flush_rows = flush_rows
        self._compaction_runs = compaction_runs
        self._key_columns = key_columns or ["id"]
        self._seqno = 0
        # writer-private working set (only the ingest thread touches these):
        self._memtable: list[pl.DataFrame] = []
        self._memtable_rows = 0
        self._runs: list[pl.DataFrame] = []
        self._snapshot = _Snapshot(runs=(), memtable=())

    def ingest(self, batch: pa.RecordBatch) -> None:
        frame = pl.from_arrow(batch)
        n = frame.height
        frame = frame.with_columns(
            pl.Series(ORDER_COLUMN, range(self._seqno, self._seqno + n))
        )
        self._seqno += n
        self._memtable.append(frame)
        self._memtable_rows += n
        if self._memtable_rows >= self._flush_rows:
            self._flush()
            if len(self._runs) >= self._compaction_runs:
                self._compact()
        self._publish()

    def _flush(self) -> None:
        if not self._memtable:
            return
        self._runs.append(pl.concat(self._memtable, how="vertical"))
        self._memtable = []
        self._memtable_rows = 0

    def _compact(self) -> None:
        merged = _merge_frame(tuple(self._runs), self._key_columns)
        if merged is None:
            self._runs = []
            return
        # Full compaction collapses every run into one, so a delete tombstone has
        # finished shadowing older versions and can be reclaimed here. A later
        # re-insert of the same key arrives with a higher seqno and wins anyway.
        merged = merged.filter(pl.col("op") != "delete")
        self._runs = [merged] if merged.height > 0 else []

    def _publish(self) -> None:
        self._snapshot = _Snapshot(runs=tuple(self._runs),
                                   memtable=tuple(self._memtable))

    def query(self, limit: int) -> tuple[list[dict], int]:
        snap = self._snapshot  # atomic read, no lock
        return _merge_to_rows(snap.runs + snap.memtable, self._key_columns, limit)

File: java-webservice-template/src/main/java/com/example/template/persistence/streamstore/LsmStore.java
package com.example.template.persistence.streamstore;

import java.util.ArrayList;
import java.util.List;

/**
 * Simplified append-only stream store.
 *
 * <p>SIMPLIFIED FROM THE PYTHON TEMPLATE: the polars window-function merge /
 * compaction is removed. Rows are appended with a monotonic seqno and returned
 * as-is (duplicates and delete tombstones included); the client receiving the
 * data performs compaction.
 *
 * <p>SINGLE-WRITER CONTRACT: exactly one thread (the ingest consumer thread)
 * calls {@link #ingest}. Readers ({@link #query}) are lock-free: each append
 * publishes a new immutable snapshot via the {@code volatile} reference, so a
 * reader always sees a consistent list.
 */
public final class LsmStore {

    public record QueryResult(List<LsmRow> rows, long total) {}

    private volatile List<LsmRow> snapshot = List.of();
    private long seqno = 0; // writer-private

    public void ingest(List<LsmRow> batch) {
        List<LsmRow> next = new ArrayList<>(snapshot);
        for (LsmRow row : batch) {
            next.add(new LsmRow(row.id(), row.name(), row.value(), seqno++, row.op()));
        }
        snapshot = List.copyOf(next); // atomic publish of immutable snapshot
    }

    public QueryResult query(int limit) {
        List<LsmRow> snap = snapshot; // atomic read
        long total = snap.size();
        List<LsmRow> rows = limit < snap.size() ? List.copyOf(snap.subList(0, limit)) : snap;
        return new QueryResult(rows, total);
    }
}
